src/sinks/gcp/cloud_storage.rs

```diff
@@ -0,0 +1,543 @@
+            retries::{RetryAction, RetryLogic},
+            tls::{TlsOptions, TlsSettings},
+            BatchBytesConfig, Buffer, PartitionBuffer, PartitionInnerBuffer, ServiceBuilderExt,
+            SinkExt, TowerRequestConfig,
+        },
+        Healthcheck, RouterSink,
+    },
+    template::Template,
+    topology::config::{DataType, SinkConfig, SinkContext, SinkDescription},
+};
+use bytes::Bytes;
+use chrono::Utc;
+use futures::{stream::iter_ok, Future, Poll, Sink};
+use http::{Method, StatusCode, Uri};
+use hyper::{
+    header::{HeaderName, HeaderValue},
+    Body, Request,
+};
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::collections::HashMap;
+use tower::{Service, ServiceBuilder};
+use tracing::field;
+use uuid::Uuid;
+
+const NAME: &str = "gcp_cloud_storage";
+const BASE_URL: &str = "https://storage.googleapis.com/";
+
+#[derive(Clone)]
+struct GcsSink {
+    bucket: String,
+    client: HttpsClient,
+    creds: Option<GcpCredentials>,
+    base_url: String,
+    settings: RequestSettings,
+}
+
+#[derive(Debug, Snafu)]
+enum GcsError {
+    #[snafu(display("Bucket {:?} not found", bucket))]
+    BucketNotFound { bucket: String },
+}
+
+#[derive(Deserialize, Serialize, Debug, Default)]
+#[serde(deny_unknown_fields)]
+pub struct GcsSinkConfig {
+    bucket: String,
+    acl: Option<GcsPredefinedAcl>,
+    storage_class: Option<GcsStorageClass>,
+    tags: Option<HashMap<String, String>>,
+    pub key_prefix: Option<String>,
+    pub filename_time_format: Option<String>,
+    pub filename_append_uuid: Option<bool>,
+    pub filename_extension: Option<String>,
+    encoding: Encoding,
+    pub compression: Compression,
+    #[serde(default)]
+    pub batch: BatchBytesConfig,
+    #[serde(default)]
+    pub request: TowerRequestConfig,
+    #[serde(flatten)]
+    pub auth: GcpAuthConfig,
+    pub tls: Option<TlsOptions>,
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+#[serde(rename_all = "camelCase")]
+pub enum GcsPredefinedAcl {
+    AuthenticatedRead,
+    BucketOwnerFullControl,
+    BucketOwnerRead,
+    Private,
+    #[derivative(Default)]
+    ProjectPrivate,
+    PublicRead,
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum GcsStorageClass {
+    #[derivative(Default)]
+    Standard,
+    Nearline,
+    Coldline,
+    Archive,
+}
+
+lazy_static! {
+    static ref REQUEST_DEFAULTS: TowerRequestConfig = TowerRequestConfig {
+        in_flight_limit: Some(25),
+        rate_limit_num: Some(25),
+        ..Default::default()
+    };
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Copy, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum Encoding {
+    #[derivative(Default)]
+    Text,
+    Ndjson,
+}
+
+impl Encoding {
+    fn content_type(self) -> &'static str {
```

Comment:
minor nit: but I think this makes more sense to take a `&self` and same for the two compression fn below.

Reply:
Is that really significant for `Copy` types, or still relevant because `&self` has a different connotation of not consuming the type?

Reply:
I personally, find it easier to read since moving to me signifies something more significant. Since we are returning a reference to me it makes sense to borrow as well. I don't mind either way just something that stood out.

Reply:
Fixed.